nano/node/repcrawler.cpp

```diff
@@ -0,0 +1,245 @@
+
+	// A representative must respond with a vote within the deadline
+	std::weak_ptr<nano::node> node_w (node.shared ());
+	node.alarm.add (std::chrono::steady_clock::now () + std::chrono::seconds (5), [node_w, hash]() {
+		if (auto node_l = node_w.lock ())
+		{
+			node_l->rep_crawler.remove (hash);
+		}
+	});
+}
+
+void nano::rep_crawler::query (nano::endpoint const & endpoint_a)
+{
+	std::vector<nano::endpoint> peers;
+	peers.push_back (endpoint_a);
+	query (peers);
+}
+
+bool nano::rep_crawler::response (nano::endpoint const & endpoint_a, nano::account const & rep_account_a, nano::amount const & weight_a)
+{
+	assert (endpoint_a.address ().is_v6 ());
+	auto updated (false);
+	std::lock_guard<std::mutex> lock (probable_reps_mutex);
+	auto existing (probable_reps.find (rep_account_a));
+	if (existing != probable_reps.end ())
+	{
+		probable_reps.modify (existing, [weight_a, &updated, rep_account_a, endpoint_a](nano::representative & info) {
+			info.last_response = std::chrono::steady_clock::now ();
+
+			if (info.weight < weight_a)
+			{
+				updated = true;
+				info.weight = weight_a;
+				info.endpoint = endpoint_a;
+				info.account = rep_account_a;
+			}
+		});
+	}
+	else
+	{
+		probable_reps.insert (nano::representative (rep_account_a, weight_a, endpoint_a));
+	}
+	return updated;
+}
+
+nano::uint128_t nano::rep_crawler::total_weight_internal ()
+{
+	nano::uint128_t result (0);
+	for (auto i (probable_reps.get<nano::tag_weight> ().begin ()), n (probable_reps.get<nano::tag_weight> ().end ()); i != n; ++i)
+	{
+		auto weight (i->weight.number ());
+		if (weight > 0)
+		{
+			result = result + weight;
+		}
+		else
+		{
+			break;
+		}
+	}
+	return result;
+}
+
+nano::uint128_t nano::rep_crawler::total_weight ()
+{
+	std::lock_guard<std::mutex> lock (probable_reps_mutex);
+	return total_weight_internal ();
+}
+
+std::vector<nano::representative> nano::rep_crawler::representatives_by_weight ()
+{
+	std::vector<nano::representative> result;
+	std::lock_guard<std::mutex> lock (probable_reps_mutex);
+	for (auto i (probable_reps.get<nano::tag_weight> ().begin ()), n (probable_reps.get<nano::tag_weight> ().end ()); i != n; ++i)
+	{
+		auto weight (i->weight.number ());
+		if (weight > 0)
+		{
+			result.push_back (*i);
+		}
+		else
+		{
+			break;
+		}
+	}
+	return result;
+}
+
+void nano::rep_crawler::on_rep_request (nano::endpoint const & endpoint_a)
+{
+	std::lock_guard<std::mutex> lock (probable_reps_mutex);
+
+	using PROBABLE_REP_ITR = PROBABLE_REP_TYPE::index<nano::tag_endpoint>::type::iterator;
+	PROBABLE_REP_TYPE::index<nano::tag_endpoint>::type & endpoint_index = probable_reps.get<nano::tag_endpoint> ();
+
+	// Find and update the timestamp on all reps available on the endpoint (a single host may have multiple reps)
+	std::vector<PROBABLE_REP_ITR> view;
+	auto itr_pair = probable_reps.get<nano::tag_endpoint> ().equal_range (endpoint_a);
+	for (; itr_pair.first != itr_pair.second; itr_pair.first++)
+	{
+		view.push_back (itr_pair.first);
```

Comment:
Is storing this list to avoid modifying the container while iterating?

Reading the documentation it seems safe to do a modification while iterating since we're modifying a non-unique index
https://www.boost.org/doc/libs/1_57_0/libs/multi_index/doc/reference/ord_indices.html#modify

Reply:
Didn't know that, thanks!